esp32_ulp/__main__.py

```diff
@@ -2,11 +2,17 @@
 
 from .util import garbage_collect
 
+from .preprocess import Preprocessor
 from .assemble import Assembler
 from .link import make_binary
 garbage_collect('after import')
 
 
+def preprocess(src):
+    preprocessor = Preprocessor()
+    return preprocessor.preprocess(src)
+
+
```

Comment:
move this to preprocess module?

Reply:
I originally had this inline further down as `src = Preprocessor().preprocess(src)` but then extracted it into a method, similar to how src_to_binary is it's own method, and knowing we'd later want to "configure" the Preprocessor instance (see commit d770b98).

It feels a bit weird to move the method, which creates instances of Preprocessor into the preprocess module (i.e. the outside world could call that method many times, creating new instances of Preprocessor all the time, without reusing the instance (that is, unless instantiating Preprocessor outside the module anyway). The singleton pattern could be an option, that the module itself will reuse an existing instance, but since the Preprocessor is a stateful instance, this also feels weird).

What was your main reason for proposing the move?

Reply:
well, it is a convenience wrapper and fits there better than in main module.
a lot of python stdlib modules offer some classes but also such convenience wrappers for the usual simple use cases.

Reply:
I added a convenience wrapper now. Looks nicer. Also in the later commit when the DefinesDB is added, it was simple enough to just add an `use_defines_db` argument to the convenience function (defaults to `True`).